cli/internal/daemon/connector/connector.go

```diff
@@ -0,0 +1,320 @@
+				return nil, ErrDaemonNotRunning
+			}
+			// The pid file doesn't exist. Start a daemon
+			pid, err := c.startDaemon()
+			if err != nil {
+				return nil, err
+			}
+			serverPid = pid
+		} else {
+			serverPid = daemonProcess.Pid
+		}
+		if err := c.waitForSocket(); errors.Is(err, ErrFailedToStart) {
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+		client, err := c.getClientConn()
+		if err != nil {
+			return nil, err
+		}
+		if err := c.sendHello(ctx, client); err == nil {
+			// We connected and negotiated a version, we're all set
+			return client, nil
+		} else if errors.Is(err, errVersionMismatch) {
+			// killLiveServer is responsible for closing the client
+			if err := c.killLiveServer(ctx, client, serverPid); err != nil {
+				return nil, err
+			}
+		} else if errors.Is(err, errConnectionFailure) {
+			// close the client, see if we can kill the stale daemon
+			_ = client.Close()
+			if err := c.killDeadServer(serverPid); err != nil {
+				return nil, err
+			}
+		} else if err != nil {
+			// Some other error occurred, close the client and
+			// report the error to the user
+			_ = client.Close()
+			return nil, err
+		}
+	}
+	return nil, ErrTooManyAttempts
+}
+
+func (c *Connector) getClientConn() (*Client, error) {
+	creds := insecure.NewCredentials()
+	conn, err := grpc.Dial(c.addr(), grpc.WithTransportCredentials(creds))
+	if err != nil {
+		return nil, err
+	}
+	tc := turbodprotocol.NewTurbodClient(conn)
+	return &Client{
+		TurbodClient: tc,
+		ClientConn:   conn,
+		SockPath:     c.SockPath,
+		PidPath:      c.PidPath,
+		LogPath:      c.LogPath,
+	}, nil
+}
+
+func (c *Connector) sendHello(ctx context.Context, client turbodprotocol.TurbodClient) error {
+	_, err := client.Hello(ctx, &turbodprotocol.HelloRequest{
+		Version: c.TurboVersion,
+		// TODO: add session id
+	})
+	status := status.Convert(err)
+	switch status.Code() {
+	case codes.OK:
+		return nil
+	case codes.FailedPrecondition:
+		return errVersionMismatch
+	case codes.Unavailable:
+		return errConnectionFailure
+	default:
+		return err
+	}
+}
+
+// waitForSocket waits for the unix domain socket to appear
+func (c *Connector) waitForSocket() error {
+	// Note that we don't care if this is our daemon
+	// or not. We started a process, but someone else could beat
+	// use to listening. That's fine, we'll check the version
+	// later.
+	deadline := time.After(_socketPollTimeout)
+	for !c.SockPath.FileExists() {
+		select {
+		case <-time.After(_socketPollInterval):
+		case <-deadline:
+			return ErrFailedToStart
+		}
+	}
+	return nil
+}
+
+// startDaemon starts the daemon and returns the pid for the new process
+func (c *Connector) startDaemon() (int, error) {
+	args := []string{"daemon"}
+	if c.Opts.ServerTimeout != 0 {
+		args = append(args, fmt.Sprintf("--idle-time=%v", c.Opts.ServerTimeout.String()))
+	}
+	c.Logger.Debug(fmt.Sprintf("starting turbod binary %v", c.Bin))
+	cmd := exec.Command(c.Bin, args...)
+	err := cmd.Start()
```

Comment:
Yup, good call, we should be setting a new pgid here

Reply:
`pgid` is now set properly, I think, so I'm resolving this half of it. I'll leave open the note on `process.Kill` since we don't do the right thing there yet.